File: src/parser.rs
use crate::*;
use std::collections::HashMap;

#[repr(u8)]
#[derive(Clone)]
pub enum BP {
    Zero,
    KeywordExpr, //trace, return, throw, if, etc.
    Assignment,
    Or,
    And,
    BitOr,
    BitAnd,
    Equality,
    Comparison,
    Addition,
    Multitplication, //(and mod)
    Exponentiation,
    Unary,
    FnCall,
    Atom
}

#[derive(Clone)]
pub struct ParserRule {
    prefix: Option<PrefixFn>,
    infix: Option<InfixFn>,
    bp: u8
}

type PrefixFn = fn(&mut Parser) -> Expr;
type InfixFn = fn(&mut Parser, Expr) -> Expr;

pub struct Parser {
    toks: TokenStream,
    pub had_error: bool,
    panic: bool,
    rules: HashMap<TokenType, ParserRule>
}

// utils
impl Parser {
    pub fn new(toks: TokenStream) -> Parser {
        let rules: HashMap<TokenType, ParserRule> = HashMap::new();
        let mut parser = Parser {
            toks,
            had_error: false,
            rules,
            panic: false
        };

        parser.rules.insert(TokenType::LeftParen,
            ParserRule {prefix: Some(Parser::grouping), infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::Minus,
            ParserRule {prefix: Some(Parser::unary),    infix: Some(Parser::binary),    bp: BP::Addition as u8});

        parser.rules.insert(TokenType::Plus,
            ParserRule {prefix: Some(Parser::unary),    infix: Some(Parser::binary),    bp: BP::Addition as u8});

        parser.rules.insert(TokenType::Slash,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Multitplication as u8});

        parser.rules.insert(TokenType::Star,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Multitplication as u8});

        parser.rules.insert(TokenType::Bang,
            ParserRule {prefix: Some(Parser::unary),    infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::Number,
            ParserRule {prefix: Some(Parser::literal),  infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::String,
            ParserRule {prefix: Some(Parser::literal),  infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::True,
            ParserRule {prefix: Some(Parser::literal),  infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::False,
            ParserRule {prefix: Some(Parser::literal),  infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::Null,
            ParserRule {prefix: Some(Parser::literal),  infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::EqualEqual,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Equality as u8});

        parser.rules.insert(TokenType::BangEqual,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Equality as u8});

        parser.rules.insert(TokenType::Greater,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Comparison as u8});

        parser.rules.insert(TokenType::GreaterEqual,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Comparison as u8});

        parser.rules.insert(TokenType::Less,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Comparison as u8});

        parser.rules.insert(TokenType::LessEqual,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Comparison as u8});

        parser.rules.insert(TokenType::Trace,
            ParserRule {prefix: Some(Parser::unarykw),  infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::LeftBrace,
            ParserRule {prefix: Some(Parser::block),    infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::Identifier,
            ParserRule {prefix: Some(Parser::access),   infix: None,                    bp: BP::Zero as u8});

        parser.rules.insert(TokenType::Equal,
            ParserRule {prefix: None,                   infix: Some(Parser::binary),    bp: BP::Assignment as u8});

        parser.rules.insert(TokenType::If,
            ParserRule {prefix: Some(Parser::if_else),  infix: None,                    bp: BP::Assignment as u8});

        parser
    }
    
    fn get_rule(&self, t: TokenType) -> ParserRule {
        let default = ParserRule {prefix: None, infix: None, bp: BP::Zero as u8};
        self.rules.get(&t).unwrap_or(&default).clone()
    }

    fn consume(&mut self, t: TokenType, msg: &'static str) -> bool {
        let cur = self.toks.peek().clone();
        if cur.t == t {
            self.toks.next();
            return true;
        }
        self.error(cur.clone(), msg);
        false
    }

    fn error(&mut self, t: Token, msg: &'static str) {
        if (self.panic) {return};
        self.had_error = true;
        self.panic = true;
        error(t, msg);
    }

    fn sync(&mut self) {
        if self.panic {
            self.panic = false;
            while self.toks.peek().t != TokenType::EOF {
                match self.toks.next().t {
                    TokenType::Semi
                    | TokenType::Class
                    | TokenType::Fn
                    | TokenType::Return
                    | TokenType::Let
                    | TokenType::Mut
                    => return,

                    _ => {}
                }
            }
        }
    }
}

// EXPR
impl Parser {
    fn parse_bp(&mut self, bp: u8) -> Expr {
        let cur = self.toks.peek();
        let prefix_rule = self.get_rule(cur.t).prefix;
        
        if let Some(f) = prefix_rule {
            let mut left = f(self);
            loop {
                let cur = self.toks.peek().clone();
                if bp > self.get_rule(cur.t).bp {break;}

                let infix_rule = self.get_rule(cur.t).infix;
                left = infix_rule.expect("This is an error with the compiler itself")(self, left);
            }
            return left;
        }

        let cur = cur.clone();
        // if there is no prefix rule for the current token, emit this error
        self.error(cur, "Expected an expression");
        return Expr::Error;
    }

    pub fn expr(&mut self) -> Expr {
        self.parse_bp(BP::Assignment as u8)
    }

    fn unary(&mut self) -> Expr {
        let op = self.toks.next();
        let expr = self.parse_bp(BP::Unary as u8);
        Expr::Unary(op, Box::new(expr))
    }

    fn unarykw(&mut self) -> Expr {
        let op = self.toks.next();
        let expr = self.parse_bp(BP::KeywordExpr as u8);
        Expr::Unary(op, Box::new(expr))
    }

    fn binary(&mut self, left: Expr) -> Expr {
        let op = self.toks.next().clone();
        
        let rule = self.get_rule(op.t);
        let right = self.parse_bp(rule.bp + 1);

        Expr::Binary(Box::new(left), op, Box::new(right))
    }

    fn literal(&mut self) -> Expr {
        let tok = self.toks.next().clone();
        match tok.t {
            TokenType::Number => Expr::Constant(tok),
            TokenType::String => Expr::Constant(tok),
            TokenType::True   => Expr::TrueLiteral(tok),
            TokenType::False  => Expr::FalseLiteral(tok),
            TokenType::Null   => Expr::NullLiteral(tok),
            _ => panic!("This is a problem with the parser itself.")
        }
    }

    fn grouping(&mut self) -> Expr {
        self.toks.next();
        let expr = self.expr();
        self.consume(TokenType::RightParen, "Unmatched (");
        expr
    }

    fn block(&mut self) -> Expr {
        self.toks.next();
        let mut stmts = Vec::<Stmt>::new();
        while self.toks.peek().t != TokenType::RightBrace && self.toks.peek().t != TokenType::EOF {
            stmts.push(self.stmt());
        }
        self.consume(TokenType::RightBrace, "Unmatched {");
        Expr::Block(stmts)
    }

    fn access(&mut self) -> Expr {
        let cur = self.toks.next();
        Expr::Access(cur.clone())
    }

    fn if_else(&mut self) -> Expr {
        self.toks.next();
        let cond = self.expr();
        self.consume(TokenType::Arrow, "Expected an arrow after the condition");
        let if_branch = self.expr();
        let mut else_branch = None;
        if self.toks.nextif(TokenType::Else) {
            else_branch = Some(Box::new(self.expr()));
        }
        Expr::IfElse(Box::new(cond), Box::new(if_branch), else_branch)
    }
}

// STMT
impl Parser {
    pub fn program(&mut self) -> Program {
        let mut stmts = Vec::<Stmt>::new();
        while self.toks.peek().t != TokenType::EOF {
            stmts.push(self.stmt());
        }
        stmts
    }
    fn stmt(&mut self) -> Stmt {
        let stmt: Stmt;
        match self.toks.peek().t {
            TokenType::Mut => stmt = self.mut_decl(),
            _ => stmt = self.expr_stmt()
        }

        if self.consume(TokenType::Semi, "Expected a SEMI after expression") {
            self.panic = false;
        }
        else {self.sync()};
        stmt
    }
    fn expr_stmt(&mut self) -> Stmt {
        let stmt = Stmt::Expr(Box::new(self.expr()));
        stmt
    }
    fn mut_decl(&mut self) -> Stmt {
        self.toks.next();
        let name = self.toks.peek().clone();
        self.consume(TokenType::Identifier, "Expected an identifier");

        let mut val = Expr::NullLiteral(name.clone());
        if self.toks.nextif(TokenType::Equal) {
            val = self.expr();
        }

        Stmt::MutDecl(name, Box::new(val))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
}

File: src/parser/Parser.ts
import { ErrorReporter } from '../error/ErrorReporter';
import { TokenType, Token } from '../lexer/Token';
import { TokenStream } from '../lexer/TokenStream';
import { noop } from '../utils';
import { match } from 'ts-pattern';
import { createASTExpr, createASTStmt, Expr, Program, Stmt } from './AST';

enum BP {
  Zero = 0,
  KeywordExpr,
  Assignment,
  ConditionalTernary,
  Or,
  And,
  Equality,
  Comparison,
  Addition,
  Multiplication,
  Exponentiation,
  Unary,
  FnCall,
  Atom,
}

type PrefixFn = () => Expr;
type InfixFn = (expr: Expr) => Expr;
type PostfixFn = (expr: Expr) => Expr;

type ParserRule = {
  prefix?: {
    fn: PrefixFn;
    bp: BP;
  };
  infix?: {
    fn: InfixFn;
    assoc: 'left' | 'right';
    bp: BP;
  };
  postfix?: {
    fn: PostfixFn;
    bp: BP;
  };
};

class Backtrack extends Error {}

const getBpForInfixRule = (
  infixRule:
    | {
        fn: InfixFn;
        assoc: 'left' | 'right';
        bp: BP;
      }
    | undefined,
) =>
  infixRule
    ? infixRule.assoc === 'left'
      ? [infixRule.bp, infixRule.bp + 0.1]
      : [infixRule.bp + 0.1, infixRule.bp]
    : [0, 0];

export class Parser {
  hadError = false;
  private panic = false;

  constructor(
    public tokenStream: TokenStream,
    public errorReporter: ErrorReporter,
  ) {}

  parse = () => this.program();

  private consume = (type: TokenType, msg: string) => {
    const cur = this.tokenStream.peek();
    if (cur.type === type) {
      this.tokenStream.next();
      return true;
    }
    this.error(cur, msg);
    return false;
  };

  private error = (token: Token, msg: string) => {
    if (this.panic || token.type === 'EOF') return;
    // console.log('error on token', token);
    this.hadError = true;
    this.panic = true;
    // console.log(
    //   `[${token.pos.from[0]}:${token.pos.from[1]}-${token.pos.to[0]}:${token.pos.to[1]}] ${msg}`,
    // );
    this.errorReporter.report(msg, token.pos);
  };

  private sync = () => {
    if (this.panic) {
      this.panic = false;
      while (this.tokenStream.peek().type !== 'EOF') {
        if (this.tokenStream.next().type === 'Semi') return;
      }
    }
  };

  private parseBp = (minBp: BP): Expr => {
    const cur = this.tokenStream.peek();
    const prefixRule = this.getRule(cur.type).prefix;

    if (prefixRule) {
      let expr = prefixRule.fn();

      while (true) {
        const cur = this.tokenStream.peek();
        const rule = this.getRule(cur.type);

        const postfixRule = rule.postfix;
        if (postfixRule) {
          const postfixBp = postfixRule.bp;
          if (postfixBp < minBp) break;

          expr = postfixRule.fn(expr);
        } else {
          const infixRule = rule.infix;
          const [lBp, rBp] = getBpForInfixRule(infixRule);
          if (lBp < minBp) break;

          expr = infixRule!.fn(expr);
        }
      }

      return expr;
    }

    this.error(cur, 'Expected an expression');
    return createASTExpr('Error');
  };

  private expr = (): Expr => {
    return this.parseBp(BP.Assignment);
  };

  private unary = (): Expr => {
    const op = this.tokenStream.next();
    const expr = this.parseBp(this.getRule(op.type).prefix!.bp);
    return createASTExpr('Unary', [op, expr]);
  };

  private unaryKeyword = (): Expr => {
    const op = this.tokenStream.next();
    const expr = this.parseBp(BP.KeywordExpr);
    return createASTExpr('Unary', [op, expr]);
  };

  private binary = (left: Expr): Expr => {
    const op = this.tokenStream.next();

    const infixRule = this.getRule(op.type).infix;
    const [lBp, rBp] = getBpForInfixRule(infixRule);
    const right = this.parseBp(rBp);

    return createASTExpr('Binary', [left, op, right]);
  };

  private conditionalTernary = (left: Expr): Expr => {
    // this is always of type Question
    const op = this.tokenStream.next();
    // console.log({ op });

    const thenExpr = this.expr();

    let elseExpr = createASTExpr('NullLiteral', [op]);
    if (this.tokenStream.nextIf('Colon')) {
      const infixRule = this.getRule(op.type).infix;
      const [lBp, rBp] = getBpForInfixRule(infixRule);
      elseExpr = this.parseBp(rBp);
    }

    // console.dir({ left, thenExpr, elseExpr }, { depth: null });

    return createASTExpr('IfElse', [left, thenExpr, elseExpr]);
  };

  private fnCall = (left: Expr): Expr => {
    // this is always of type LeftParen
    const op = this.tokenStream.next();

    let args: Expr[] = [];

    if (this.tokenStream.peek().type !== 'RightParen') {
      while (true) {
        args.push(this.expr());

        if (!this.tokenStream.nextIf('Comma')) {
          break;
        }
      }
    }

    this.consume('RightParen', 'Expected ) after argument list');

    return createASTExpr('FnCall', [left, args]);
  };

  private literal = (): Expr => {
    const tok = this.tokenStream.next();

    const expr = match(tok.type)
      .with('Number', () => createASTExpr('Constant', [tok]))
      .with('String', () => createASTExpr('Constant', [tok]))
      .with('True', () => createASTExpr('TrueLiteral', [tok]))
      .with('False', () => createASTExpr('FalseLiteral', [tok]))
      .with('Null', () => createASTExpr('NullLiteral', [tok]))
      .otherwise(() => {
        throw new Error('This is an error with the parser itself!');
      });

    return expr;
  };

  private grouping = (): Expr => {
    this.tokenStream.next();

    const backtrackPoint = this.tokenStream.backtrackPoint();
    try {
      const expr = this.fnFromLeftParen();
      return expr;
    } catch (err) {
      // console.log(err);
      backtrackPoint();

      // console.log(this.tokenStream.peek());

      const expr = this.expr();

      this.consume('RightParen', 'Unmatched (');

      return expr;
    }
  };

  private identifier = (): Expr => {
    const cur = this.tokenStream.next();

    if (this.tokenStream.peek().type === 'SlimArrow') {
      return this.fnFromIdentifier(cur);
    }

    return createASTExpr('Access', [cur]);
  };

  private program = (): Program => {
    const stmts: Program = [];
    let isFirst = true;
    while (this.tokenStream.peek().type !== 'EOF') {
      stmts.push(this.stmt(isFirst));
      isFirst = false;
    }

    return stmts;
  };

  private stmt = (isFirst: boolean): Stmt => {
    if (!isFirst) {
      if (this.consume('Semi', 'Expected a ; after expression')) {
        this.panic = false;
      } else {
        this.sync();
      }
    }

    const tok = this.tokenStream.peek();

    match(tok.type)
      .with('RightParen', () => {
        this.error(
          tok,
          'Unexpected semi. Either remove the semi, or explicitly return null',
        );
      })
      .with('EOF', () => {
        this.error(
          tok,
          'The last statement in your program must not end in a semi',
        );
      });

    return match(tok.type)
      .with('Mut', () => this.mutDecl())
      .otherwise(() => this.exprStmt());
  };

  private exprStmt = (): Stmt => {
    return createASTStmt('Expr', [this.expr()]);
  };

  private fnParamList = (endWith: 'RightParen' | 'SlimArrow'): Token[] => {
    let params: Token[] = [];
    if (this.tokenStream.peek().type !== endWith) {
      while (true) {
        params.push(this.tokenStream.peek());
        this.consume('Identifier', 'Expected parameter name');

        if (!this.tokenStream.nextIf('Comma')) {
          break;
        }
      }
    }
    return params;
  };

  private fnFromLeftParen = (): Expr => {
    const backtrackPoint = this.tokenStream.backtrackPoint();

    // in this loop we detect if this statement has a ") ->" somewhere
    // if so, we backtrack and parse the argument list and the whole lambda
    // otherwise we throw and backtrack to try and parse a grouping instead

    // TODO: limit this reasonably to only several tokens
    while (true) {
      // look as far as a semi, eof, or leftparen
      // the LeftParen comes from the fact, that we
      // might have an expression like (a + (b) -> c)
      // where the existence of the nested lambda
      // confuses the parser, which now tries to parse the grouping as a lambda
      if (['Semi', 'EOF', 'LeftParen'].includes(this.tokenStream.peek().type)) {
        throw new Backtrack();
      }
      if (this.tokenStream.peek().type === 'RightParen') {
        this.tokenStream.next();
        if (this.tokenStream.peek().type !== 'SlimArrow') {
          throw new Backtrack();
        }
        this.tokenStream.next();
        break;
      }
      this.tokenStream.next();
    }
    backtrackPoint();

    const params = this.fnParamList('RightParen');

    this.tokenStream.next(); // RightParen
    this.tokenStream.next(); // SlimArrow

    return createASTExpr('Fn', [params, this.expr()]);
  };

  private fnNoParams = (): Expr => {
    this.tokenStream.next();
    return createASTExpr('Fn', [[], this.expr()]);
  };

  private fnFromIdentifier = (name: Token): Expr => {
    this.tokenStream.next();
    return createASTExpr('Fn', [[name], this.expr()]);
  };

  private mutDecl = (): Stmt => {
    this.tokenStream.next();

    const name = this.tokenStream.peek();
    this.consume('Identifier', 'Expected a variable identifier');

    let val = createASTExpr('NullLiteral', [name]);
    if (this.tokenStream.nextIf('Equal')) {
      val = this.expr();
    }

    return createASTStmt('MutDecl', [name, val]);
  };

  private rules: { [key in TokenType]?: ParserRule } = {
    LeftParen: {
      prefix: {
        fn: this.grouping,
        bp: BP.Unary,
      },
      postfix: {
        fn: this.fnCall,
        bp: BP.FnCall,
      },
    },
    Plus: {
      prefix: { fn: this.unary, bp: BP.Unary },
      infix: { fn: this.binary, bp: BP.Addition, assoc: 'left' },
    },
    Minus: {
      prefix: { fn: this.unary, bp: BP.Unary },
      infix: { fn: this.binary, bp: BP.Addition, assoc: 'left' },
    },
    Slash: {
      infix: { fn: this.binary, bp: BP.Multiplication, assoc: 'left' },
    },
    Star: {
      infix: { fn: this.binary, bp: BP.Multiplication, assoc: 'left' },
    },
    Bang: {
      prefix: { fn: this.unary, bp: BP.Unary },
    },

    Number: {
      prefix: { fn: this.literal, bp: BP.Atom },
    },
    String: {
      prefix: { fn: this.literal, bp: BP.Atom },
    },
    True: {
      prefix: { fn: this.literal, bp: BP.Atom },
    },
    False: {
      prefix: { fn: this.literal, bp: BP.Atom },
    },
    Null: {
      prefix: { fn: this.literal, bp: BP.Atom },
    },

    EqualEqual: {
      infix: { fn: this.binary, bp: BP.Equality, assoc: 'left' },
    },
    BangEqual: {
      infix: { fn: this.binary, bp: BP.Equality, assoc: 'left' },
    },

    Greater: {
      infix: { fn: this.binary, bp: BP.Comparison, assoc: 'left' },
    },
    GreaterEqual: {
      infix: { fn: this.binary, bp: BP.Comparison, assoc: 'left' },
    },
    Less: {
      infix: { fn: this.binary, bp: BP.Comparison, assoc: 'left' },
    },
    LessEqual: {
      infix: { fn: this.binary, bp: BP.Comparison, assoc: 'left' },
    },

    Identifier: {
      prefix: { fn: this.identifier, bp: BP.Atom },
    },

    And: {
      infix: { fn: this.binary, bp: BP.And, assoc: 'left' },
    },
    Or: {
      infix: { fn: this.binary, bp: BP.Or, assoc: 'left' },
    },

    SlimArrow: {
      prefix: { fn: this.fnNoParams, bp: BP.Unary },
    },

    Question: {
      infix: {
        fn: this.conditionalTernary,
        bp: BP.ConditionalTernary,
        assoc: 'right',
      },
    },
  };

  private getRule = (type: TokenType) => {
    const rule = this.rules[type];

    return rule ?? {};
  };
}
